Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -6,102 +6,195 @@
-    }
+// MARK: - Private
 
-    static let subjectClass = URLSession.self
-    private static let swizzler = MethodSwizzler.shared
+private let swizzler = MethodSwizzler.shared
 
-    static func swizzle() throws {
-        try swizzleDataTaskWithURL()
-        try swizzleDataTaskWithRequest()
-        try swizzleDataTaskWithURLCompletionHandler()
-        try swizzleDataTaskWithRequestCompletionHandler()
-    }
+internal enum URLSessionSwizzler {
+    static var hasSwizzledBefore = false
 
-    static func unswizzle() throws {
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithURL, in: subjectClass)
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithRequest, in: subjectClass)
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithURLCompletion, in: subjectClass)
-        try swizzler.unswizzle(selector: Selectors.DataTaskWithRequestCompletion, in: subjectClass)
-    }
+    static func swizzleOnce(using interceptor: @escaping RequestInterceptor) throws {
+        guard let method_dataTask_URL = DataTask_URL_Completion.foundMethod,
+            let method_dataTask_request = DataTask_Request_Completion.foundMethod else {
+                throw InternalError(description: "URLSession methods could not be found, thus not swizzled")
+        }
 
-    static func swizzleDataTaskWithURL() throws {
-        // typealiases cannot be generic as C/block conventions don't support generics
-        // note that _Block doesn't have Selector parameter
-        typealias TypedIMP = TypedIMPs.DataTaskWithURL
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithURL
-        let sel = Selectors.DataTaskWithURL
+        if hasSwizzledBefore {
+            return
+        }
 
-        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+        let newImp_URL = DataTask_URL_Completion.newIMP(
+            using: interceptor,
+            methodToSwizzle: method_dataTask_URL,
+            methodToRedirect: method_dataTask_request
+        )
+        swizzler.set(newIMP: newImp_URL, for: method_dataTask_URL)
 
-        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURL -> URLSessionDataTask in
-            return typedOriginalImp(impSelf, impSelector, impURL)
-        }
-        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
-        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
-    }
+        let newImp_request = DataTask_Request_Completion.newIMP(
+            using: interceptor,
+            methodToSwizzle: method_dataTask_request
+        )
+        swizzler.set(newIMP: newImp_request, for: method_dataTask_request)
 
-    static func swizzleDataTaskWithRequest() throws {
-        typealias TypedIMP = TypedIMPs.DataTaskWithRequest
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithRequest
-        let sel = Selectors.DataTaskWithRequest
+        hasSwizzledBefore = true
+    }
```

Comment:
I found this `IMP`-dance very fragile and error-prone. In fact, I tried to mess it up randomly, and it compiled, so even if we enforce the type-safety for `IMP` blocks, we don't leverage it:

<img width="488" alt="Screenshot 2020-06-03 at 13 09 02" src="https://user-images.githubusercontent.com/2358722/83641726-33318f80-a5ae-11ea-81d7-f2d44d828c54.png">

I gave it a shot and tried if an object-oriented approach can help. Here's what I came up with:

```swift
static func swizzleOnce(using interceptor: @escaping RequestInterceptor) throws {
    if hasSwizzledBefore {
        return
    }
    defer { hasSwizzledBefore = true }

    guard let dataTaskWithURL = DataTaskWithURL(), let dataTaskWithURLRequest = DataTaskWithURLRequest() else {
        throw InternalError(description: "URLSession methods could not be found, thus not swizzled")
    }

    dataTaskWithURL.intercept(using: interceptor, byRedirectingTo: dataTaskWithURLRequest)
    dataTaskWithURLRequest.intercept(using: interceptor)
}
```

Modeling methods as objects is natural when solvign this problem in OOP. It's easy to follow and we keep a full control over what is passed where. It is impossible to use it wrong. See the code [here](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-452-propose-swizzling-refactoring/Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift).

Reply:
correct 👍 those were leaked during the last refactor yesterday, this looks better; making the change ✅

Reply:
`dataTaskWithURL/Request` swizzlers are %~100 self-contained, their types or properties cannot be mixed anymore ✅ 

also i made another change that simplifies the code:
`dataTaskWithURL` does redirection with normal method call now
```swift
let task = impSelf.dataTask(with: interceptionResult.modifiedRequest, completionHandler: modifiedCompletion)
```

as stated in commit message,
🙌 Pros: we don't swallow further swizzling on `dataTaskWithRequest`
😞 Cons: we will execute requestInterceptor twice for the same URL
wdyt?

Reply:
> we will execute requestInterceptor twice for the same URL

Sounds like this will create two spans for single request, which is not acceptable. Even if not, "this block is called twice" is definitely not what we want.

Reply:
update: this discussion was resolved in Slack. we never call `interceptor` twice ✅